File: Examples/Objective-C/BroadcastAlert/BroadcastAlert/ReceivedNotificationsViewController.m
//
//  ReceivedNotificationsViewController.m
//  BroadcastAlert
//
//  Created by Daniel Heredia on 7/27/16.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

#import "Notification.h"
#import "ReceivedNotificationsViewController.h"
#import "Constants.h"

#ifndef ReceivedNotifs
#define ReceivedNotifs
#define FULLPATH(X) [[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0] \
    stringByAppendingPathComponent:(X)]
#define kNotificationsFile @"notifs.txt"
#endif

@interface ReceivedNotificationsViewController ()
@property (nonatomic, retain) NSArray* notifications;
@end

@implementation ReceivedNotificationsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    //Load previous notifications
    self.notifications = [self.class loadNotifications];
    
    // Prevents a shadow on the nav bar when pushed
    self.navigationController.view.backgroundColor = [UIColor whiteColor];
}

- (void)addNotificationDictionary:(NSDictionary*)dictionary fromUUID:(NSString*)uuid
{
    //Process the data sent by other peer.
    [self.class addNotificationToFile:dictionary
                             fromUUID:uuid];
    [self refreshData];
}

- (void)refreshData
{
    self.notifications = [self.class loadNotifications];
    [self.tableView reloadData];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView*)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView*)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.notifications count];
}

- (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSIndexPath*)indexPath
{
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:@"notificationCell" forIndexPath:indexPath];
    Notification* notification = [self.notifications objectAtIndex:indexPath.item];
    UILabel *fromUsernameLabel = [(UILabel *)cell.contentView viewWithTag:1001];
    UILabel *fromUUIDLabel = [(UILabel *)cell.contentView viewWithTag:1002];
    UILabel *alertNumberLabel = [(UILabel *)cell.contentView viewWithTag:1003];
    UILabel *dateLabel = [(UILabel *)cell.contentView viewWithTag:1004];
    UILabel *timeLabel = [(UILabel *)cell.contentView viewWithTag:1005];
    
    fromUsernameLabel.text = notification.senderName;
    fromUUIDLabel.text = [NSString stringWithFormat:@"(%@)", notification.senderId];
    alertNumberLabel.text = [NSString stringWithFormat:@"%ld", (long)notification.number];
    
    NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"dd/MM/yyyy"];
    NSString* dateString = [dateFormatter stringFromDate:notification.date];
    dateLabel.text = dateString;
    
    [dateFormatter setDateFormat:@"HH:mm:ss"];
    NSString *timeString = [dateFormatter stringFromDate:notification.date];
    timeLabel.text = timeString;

    return cell;
}

#pragma mark - Clumsy data management

// The methods in this section are not relevant to show
// the BFTransmitter functionality.

+ (void)addNotificationToFile:(NSDictionary*)dictionary fromUUID:(NSString*)uuid
{
    Notification* notification = [[Notification alloc] init];
    notification.number = [dictionary[@"number"] integerValue];
    notification.senderId = [uuid substringToIndex:5];
    notification.senderName = dictionary[@"device_name"];
    double doubleDate = [dictionary[@"date_sent"] doubleValue] / 1000;
    NSDate* date = [NSDate dateWithTimeIntervalSince1970:doubleDate];
    notification.date = date;

    NSMutableArray* notifications = [self loadNotifications];
    [notifications insertObject:notification
                        atIndex:0];

    NSString* filePath = FULLPATH(kNotificationsFile);
    NSData* data = [NSKeyedArchiver archivedDataWithRootObject:notifications];
    [data writeToFile:filePath atomically:YES];
}

+ (NSMutableArray*)loadNotifications
{
    NSString* filePath = FULLPATH(kNotificationsFile);
    NSData* data = [NSData dataWithContentsOfFile:filePath];
    NSMutableArray* notifications;
    
    if (data) {
        notifications = [NSKeyedUnarchiver unarchiveObjectWithData:data];
    }

    if (notifications == nil)
        notifications = [[NSMutableArray alloc] init];

    return notifications;
}

+ (BOOL)clearReceivedNotifications {
    
    NSString* filePath = FULLPATH(kNotificationsFile);
    NSData* data = [NSKeyedArchiver archivedDataWithRootObject:[[NSMutableArray alloc] init]];
    return [data writeToFile:filePath atomically:YES];
}

@end

File: Examples/Swift/BroadcastAlert/BroadcastAlert/ReceivedNotificationsViewController.swift
//
//  ReceivedNotificationsViewController.swift
//  BroadcastAlert
//
//  Created by Daniel Heredia on 8/2/16.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

import UIKit

let notifsFile = "notifs.txt"

open class ReceivedNotificationsViewController: UITableViewController {
    
    fileprivate var notifications: NSMutableArray

    public required init?(coder aDecoder: NSCoder) {
        //Load previous notifications
        self.notifications = ReceivedNotificationsViewController.loadNotifications()
        super.init(coder: aDecoder)
        
    }

    open override func viewDidLoad() {
        super.viewDidLoad()
    }

    // MARK: Info update interface methods
    
    open func addNotificationDictionary(_ dictionary: [AnyHashable: Any], fromUUID uuid: String) {
            //Process the data sent by other peer.
            ReceivedNotificationsViewController.addNotificationToFile(dictionary, fromUUID: uuid)
            self.refreshData()
        }
        
    func refreshData() {
        self.notifications = ReceivedNotificationsViewController.loadNotifications()
        self.tableView.reloadData()
    }

    // MARK: Table view data source
    
    open override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    open override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.notifications.count
    }
    
    open override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "notificationCell", for: indexPath)
        let notification: Notification = self.notifications.object(at: indexPath.item) as! Notification
        let numberLabel = cell.contentView.viewWithTag(1001) as! UILabel
        let fromLabel = cell.contentView.viewWithTag(1002) as! UILabel
        let dateLabel = cell.contentView.viewWithTag(1003) as! UILabel
        numberLabel.text = "Alert number: \(notification.number)"
        fromLabel.text = "From user: \(notification.senderName) (\(notification.senderId))"
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "HH:mm:ss"
        let dateString = dateFormatter.string(from: notification.date as Date)
        dateLabel.text = "Time: \(dateString)"
        return cell
    }

    // MARK: Clumsy data management.
    
    // The methods in this section are not relevant to show
    // the BFTransmitter functionality.

    class func addNotificationToFile(_ dictionary: [AnyHashable: Any], fromUUID uuid: String) {
        let notification = Notification()
        notification.number = (dictionary["number"] as! NSNumber).intValue
        notification.senderId = uuid.substring(to: uuid.characters.index(uuid.startIndex, offsetBy: 5))
        notification.senderName = dictionary["device_name"] as! String
        let doubleValue = (dictionary["date_sent"] as! NSNumber).doubleValue / 1000
        let date = Date(timeIntervalSince1970: doubleValue)
        notification.date = date
        let notifications: NSMutableArray = self.loadNotifications()
        notifications.insert(notification, at: 0)
        let filePath = fullPathForFile(notifsFile)
        let data = NSKeyedArchiver.archivedData(withRootObject: notifications)
        try? data.write(to: URL(fileURLWithPath: filePath), options: [.atomic])
    }
    
    class func loadNotifications() -> NSMutableArray {
        let filePath = fullPathForFile(notifsFile)
        let data: Data? = try? Data(contentsOf: URL(fileURLWithPath: filePath))
        if (data != nil)
        {
            return NSKeyedUnarchiver.unarchiveObject(with: data!) as! NSMutableArray
        } else
        {
            return NSMutableArray()
        }
    }

    class func fullPathForFile(_ file: String) -> String {
        let path = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0] as String
        let url = URL(fileURLWithPath: path)
        return url.appendingPathComponent(file).path
    }
}
